added AD compatible spectral decomposition for ranktwotensor

hugary1995 opened the pull request:
derived an algorithm from EISPACK to solve eigenvalue decomposition of a ranktwotensor by using a tridiagonal reduction and a QL decomposition

refs #13191

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

hugary1995 committed:
added AD compatible spectral decomposition for ranktwotensor

added several utilities to RankTwoTensor, including Givens rotation, Hessenberg transformation and QR factorization.
An iterative eigenvalue algorithm is then derived from those utilities to solve for eigenvalues and eigenvectors of
a symmetric rank two tensor. DualReal instantiations are added when applicable. Added unit tests for Hessenberg, QR,
and eigenvalue algorithms. Dual number consistency is checked.

refs #13191

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,256 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
 
+template <typename T>
+void
+RankTwoTensorTempl<T>::adSymmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T> & Q,
```
Don't add a separate method. Instead make this a template specialization for `DualReal`

```
template <>
void
RankTwoTensorTempl<DualReal>::symmetricEigenvaluesEigenvectors(std::vector<DualReal> & eigvals, RankTwoTensorTempl<DualReal> & eigvecs) const
{
...
}
```

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -891,11 +891,11 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
 void
-RankTwoTensorTempl<T>::adSymmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T> & Q,
-                                                          RankTwoTensorTempl<T> & D) const
+RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T> & Q,
```
No, this is not a template specialization.

dschwen replied:
And why does this have a different signature than the other method?

hugary1995 replied:
The decomposition is written as A = Q D Q^T, and some of the dual numbers reside in off-diagonal entries of the diagonal matrix D. I know ideally there shouldn't be any off-diagonal dual numbers, but all practical algorithms I have tried give me non-zero off-diagonal dual numbers.

Therefore, I am requesting and returning not just std::vector<DualReal> & eigvals, but the entire eigenvalue matrix RankTwoTensorTempl<T> & D. That's why I am writing it as a function overloading rather than a template specialization right now.

dschwen commented:
A 3*3 matrix has a maximum of 3 Eigenvalues in the Math I know. What am
I missing?

On Fri, May 10, 2019, 12:07 PM Gary (Tianchen) Hu <notifications@github.com>
wrote:

> *@hugary1995* commented on this pull request.
> ------------------------------
>
> In framework/src/utils/RankTwoTensor.C
> <https://github.com/idaholab/moose/pull/13405#discussion_r282987328>:
>
> > @@ -891,11 +891,11 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
>
>  template <typename T>
>  void
> -RankTwoTensorTempl<T>::adSymmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T> & Q,
> -                                                          RankTwoTensorTempl<T> & D) const
> +RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T> & Q,
>
> The decomposition is written as A = Q D Q^T, and some of the dual numbers
> reside in off-diagonal entries of the diagonal matrix D. I know ideally
> there shouldn't be any off-diagonal dual numbers, but all practical
> algorithms I have tried give me non-zero off-diagonal dual numbers.
>
> Therefore, I am requesting and returning not just std::vector & eigvals,
> but the entire eigenvalue matrix RankTwoTensorTempl & D. That's why I am
> writing it as a function overloading rather than a template specialization
> right now.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13405#discussion_r282987328>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABRMPXNXFFFAN5QI5LBEY3PUW2VLANCNFSM4HMEGFVQ>
> .
>

dschwen commented:
And what are the eigenvectors for those "offdiagonal Eigenvalues" supposed
to be?

On Fri, May 10, 2019, 12:11 PM Daniel Schwen <notifications@github.com>
wrote:

> A 3*3 matrix has a maximum of 3 Eigenvalues in the Math I know. What am
> I missing?
>
> On Fri, May 10, 2019, 12:07 PM Gary (Tianchen) Hu <
> notifications@github.com>
> wrote:
>
> > *@hugary1995* commented on this pull request.
> > ------------------------------
> >
> > In framework/src/utils/RankTwoTensor.C
> > <https://github.com/idaholab/moose/pull/13405#discussion_r282987328>:
> >
> > > @@ -891,11 +891,11 @@
> RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> &
> eigvals
> >
> > template <typename T>
> > void
> >
> -RankTwoTensorTempl<T>::adSymmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T>
> & Q,
> > - RankTwoTensorTempl<T> & D) const
> >
> +RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(RankTwoTensorTempl<T>
> & Q,
> >
> > The decomposition is written as A = Q D Q^T, and some of the dual numbers
> > reside in off-diagonal entries of the diagonal matrix D. I know ideally
> > there shouldn't be any off-diagonal dual numbers, but all practical
> > algorithms I have tried give me non-zero off-diagonal dual numbers.
> >
> > Therefore, I am requesting and returning not just std::vector & eigvals,
> > but the entire eigenvalue matrix RankTwoTensorTempl & D. That's why I am
> > writing it as a function overloading rather than a template
> specialization
> > right now.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/13405#discussion_r282987328>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AABRMPXNXFFFAN5QI5LBEY3PUW2VLANCNFSM4HMEGFVQ
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13405#issuecomment-491382404>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABRMPWL5TOFA7467WVJSA3PUW3GFANCNFSM4HMEGFVQ>
> .
>

hugary1995 commented:
Those off-diagonal entries in the eigenvalue matrix are zero. Ideally you have three eigenvalues d1, d2, d3 and eigenvectors v1, v2, v3. And A = v1 d1 v1^T + v2 d2 v2^T + v3 d3 v3^T. If you do that in AD you won't get correct jacobian because those practical eigenvalue algorithms give you
```
D = [ d1(some dual numbers)    0(some dual numbers)     0(some dual numbers)
        0(some dual numbers)      d2(some dual numbers)    0(some dual numbers)
        0(some dual numbers)       0(some dual numbers)     d3(some dual numbers)]
```

hugary1995 commented:
off-diagonal entries in D are there so that you get consistent A = Q D Q^T not just in values, but also in their dual numbers. I know this is not strict in the Math but currently it works well for all my test problems. I don't have to merge this and we can wait until someone figures out a perfect AD-compatible eigenvalue algorithm.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/337299/) on 0bc6dce wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13405/site)

This comment will be updated on new commits.

hugary1995 commented:
@dschwen I changed it to template instantiation. Please take a look when HPC maintenance is done. Should I set up a test for this as well?

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+                                                  std::vector<T> & d,
+                                                  T e[3]) const
+{
+  int N = 3;
```
This class already has an `N` member that is set to LIBMESH_DIM (ususally 3)

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+
+  T f = 0.0;
+  T tst1 = 0.0;
+  T eps = std::pow(2.0, -52.0);
```
That is a ... new ... way to define an epsilon. This should be `const` for sure and can be of type `Real`! I suggest checking if you can use `libMesh::TOLERANCE` or `libMesh::TOLERANCE * libMesh::TOLERANCE` here.

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+    int m = l;
+    while (m < N)
+    {
+      if (std::fabs(e[m]) <= eps * tst1)
```
please use `std::abs` and not `std::fabs`

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+        T c = 1.0;
+        T c2 = c;
+        T c3 = c;
+        T el1 = e[l + 1];
```
please check which variables can be marked `const`. This makes the code more explicit and may help the compiler perform further optimizations.

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+  e[N - 1] = 0.0;
+
+  T f = 0.0;
+  T tst1 = 0.0;
```
It looks like `tst1` can be of type `Real`. It's value is only used in conditionals and never contributes to the returned values directly, thus its derivatives are not required or used.

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+void
+RankTwoTensorTempl<T>::tridiagonalQLDecomposition(RankTwoTensorTempl<T> & V,
+                                                  std::vector<T> & d,
+                                                  T e[3]) const
```
If possible let's avoid raw C arrays. You could pass in a reference to a `std::array<T, RankTwoTensorTempl<T>:: N>`.

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+    T scale = 0.0;
+    T h = 0.0;
+    for (int k = 0; k < i; k++)
+      scale = scale + std::fabs(d[k]);
```
```
scale += std::abs(d[k]);
```

dschwen commented:
@hugary1995 this looks pretty good. Thanks for making it work! I added a few small comments.

dschwen commented:
Ok, coincidentally I just ran into a situation where I really need this, too!

It would be great if you could add a unit test that generates a few (RankTwoTensor, DualRankTwoTensor) paris with identical values and compares the spectral decomposition for both.

Let me know if you need help with that.

hugary1995 commented:
No problem let me try... I was also thinking about adding a unit test.

dschwen commented:
This should be a unit test

hugary1995 commented:
@dschwen I think you can easily understand what I am doing here by looking at the input file of the test I added. For your convenience here is some brief explanation:

I added six aux variables a00, a01, a02, a11, a12, a22. Each of them is initialized by a RandomIC with a uniform distribution ~ [-100, 100].

In a material class called "ADRankTwoTensorSymmetricEigenvaluesEigenvectorsTest" I use these six auxvariables to construct a MaterialProperty<RankTwoTensor>.

In the template instantiation of ComputeStage::RESIDUAL I explicitly cast the RankTwoTensor at qp to DualRankTwoTensor, and call symmetricEigenvaluesEigenvectors for each of them.

For the RankTwoTensor I can get eigenvalues da1, da2, da3 and eigenvectors va1, va2, va3.
For the DualRankTwoTensor I can get eigenvalues db1, db2, db3 and eigenvectors vb1, vb2, vb3.

The difference in eigenvalues is declared as MaterialProperty<Real> _error_eigval
The difference in eigenvectors is declared as MaterialProperty<Real> _error_eigvec

_error_eigval = |da1 - db1| + |da3 - db3| + |da3 - db3|
_error_eigvec = ||va1 x vb1|| + ||va2 x vb2|| + ||va3 x vb3||

The mesh has 10000 elements, so there are around 40000 quadrature points hence 40000 random rank two tensors.

The maximum eigenvalue error is on the order of 10^-14.
The maximum eigenvector error is on the order of 10^-11.

dschwen commented:
Yeah, I'm very happy to sea this agreement, but this should still be a unit test.

hugary1995 commented:
Ohhhh okay, I'll change it to a unit test. Should I create a new ADRankTwoTensorTest or stay within the RankTwoTensorTest?

hugary1995 commented:
@dschwen I enabled AD for RankTwoEigenRoutines. I was surprised we didn't have a test case for symmetricEigenvaluesEigenvectors for Real RankTwoTensor before so I added a couple...

The "RankTwoEigenRoutines.symmetricEigenvaluesEigenvectors_dual_number_consistency" demonstrates current limitations of the eigenvalue algorithm. As I mentioned, dual numbers of off-diagonal entries are inconsistent...

dschwen commented on unit/src/RankTwoEigenRoutinesTest.C:
```diff
@@ -9,9 +9,50 @@
 #include "RankTwoTensor.h"
 #include "RankFourTensor.h"
 
+#define setDualReal(name, v, d0, d1, d2)                                                           \
```
Why are those macros?

hugary1995 replied:
I can remove those...

hugary1995 commented:
@dschwen Sorry I didn't let you know earlier -- I rewrote it from scratch over the holiday. Now the dual number consistency issue should be fixed but I need more testing. Do you mind me changing it to WIP and force push later?

dschwen commented:
Yeah, that sounds good to me.

hugary1995 committed:
handle a corner case when we need to permute the tensor before rotating dual numbers

hugary1995 commented:
@dschwen I think now this is ready for review. I broke the original unreadable EISPACK algorithm to several small, manageable pieces, which by themselves are very useful linear algebra operations.

The basic building blocks for the new eigenvalue algorithm are: row/column permutation, Givens rotation, QR factorization (Givens version). I instantiated them for DualReal whenever applicable.

The eigenvalue algorithm itself is a simple QR-RQ algorithm which iteratively computes the QR factorization of the tensor.

I also added several print methods for DualReal which I found useful for debugging.

@jiangwen84 Can we try it in a AD phase-field fracture material next?

jiangwen84 commented:
@hugary1995 👍  Once this gets merged, I will push my AD phase field fracture.

hugary1995 committed:
handle the very last corner case!

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -369,6 +369,12 @@ class RankTwoTensorTempl : public TensorValue<T>
   void print(std::ostream & stm = Moose::out) const;
 
+  /// Print the Real part of the DualReal rank two tensor
+  void printReal(std::ostream & stm = Moose::out) const;
```
How about we fold all this into the original print method by adding a second default parameter `unsigned int nDual = 0`?

dschwen replied:
ok, never mind, don't change anything now

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +927,263 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+                             (b.value() * sin.value() + a.value() * cos.value());
+      sin.derivatives()[i] = (b.derivatives()[i] * cos.value() * cos.value() -
+                              a.derivatives()[i] * cos.value() * sin.value()) /
+                             (b.value() * sin.value() + a.value() * cos.value());
```
why do you need to manually update the derivatives here? Why does the automatic differentiation not work?

hugary1995 replied:
I wrote this before I fixed the QR algorithm. We shouldn't need this anymore. Let me remove it.

The very issue of this entire algorithm is that when both entries to rotate are 0. There is no way to rotate the dual numbers (imagine rotating a zero-length vector), which is why I did additional permutation before doing the rotation.

hugary1995 committed:
remove unnecessary DualReal instantiation of GivensRotation

dschwen reviewed (dismissed):
Ok, let's get this in.

hugary1995 commented:
Sorry wait one sec... I have one last code optimization.

hugary1995 committed:
unify row/col permutation, remove redundant permutations in QR

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -929,29 +929,20 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
 RankTwoTensorTempl<T>
-RankTwoTensorTempl<T>::rowPermutationTensor(std::vector<unsigned int> old_row,
-                                            std::vector<unsigned int> new_row) const
+RankTwoTensorTempl<T>::permutationTensor(std::vector<unsigned int> old_elements,
```
can you pass in a const reference instead, please

dschwen replied:
maybe you can use a `TypeVector<unsigned int>` here? Or an `std::array<unsigned int, LIBMESH_DIM>`

hugary1995 replied:
I think TypeVector's constructor is private here, I'll change them to `const std::array<unsigned int, LIBMESH_DIM> &`

dschwen commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -929,29 +929,20 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
-  mooseAssert(N == old_row.size() && N == new_row.size(),
-              "rowPermutationTensor: number of rows should be equal to LIBMESH_DIM");
+  mooseAssert(N == old_elements.size() && N == new_elements.size(),
+              "permutationTensor: number of permuting elements should be equal to LIBMESH_DIM");
```
The check and the assertion message are not checking/describing the same thing here

hugary1995 committed:
address comments

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/345292/) on 5b7b665 : invalidated by @dschwen

permcody commented on framework/src/utils/RankTwoTensor.C:
```diff
@@ -889,6 +927,182 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+  this->hessenberg(D, eigvecs);
+
+  unsigned int iter = 0;
+  for (unsigned m = N - 1; m > 0; m--)
```
How did this not error? unsigned ints are always greater than zero. This is a bad loop test!

hugary1995 replied:
Hmm I learned it a long time ago so I might be mistaken -- isn't 000...0 in signed bit a zero?

permcody replied:
Sorry - you are correct

` m > 0` is valid
` m >= 0` is not valid for unsigned

permcody reviewed (commented):
@hugary1995 - Can you take a look at moosebuild.org and see if you can fix up the Clang minimum compiler errors that this PR caused?

hugary1995 commented:
@permcody No problem I'll fix it. I didn't know that older compilers prefer double braces until I read [this](https://stackoverflow.com/questions/31555584/why-is-clang-warning-suggest-braces-around-initialization-of-subobject-wmis)

permcody commented:
Thanks!
